pkg/datapath/linux/ipsec/ipsec_linux.go

```diff
@@ -620,3 +625,49 @@ func DeleteIPsecEncryptRoute() {
 		}
 	}
 }
+
+func keyfileWatcher(watcher *fsnotify.Watcher, keyfilePath string, nodediscovery *nodediscovery.NodeDiscovery) {
+	for {
+		select {
+		case event := <-watcher.Events:
+			if event.Name != keyfilePath {
+				continue
+			}
+
+			log.WithField(logfields.Path, keyfilePath).Debugf("Received %s fsnotify event for IPSec keyfile", event.Op.String())
+			if event.Op&(fsnotify.Create|fsnotify.Write) == 0 {
+				continue
+			}
+
+			_, spi, err := LoadIPSecKeysFile(keyfilePath)
+			if err != nil {
+				log.WithError(err).Errorf("Failed to load IPsec keyfile")
+			}
+
+			node.SetIPsecKeyIdentity(spi)
+			nodediscovery.UpdateNode()
+
+		case err := <-watcher.Errors:
+			log.WithError(err).WithField("path", keyfilePath).Warning("error encountered while watching file with fsnotify")
+		}
+	}
+}
+
+func StartKeyfileWatcher(keyfilePath string, nodediscovery *nodediscovery.NodeDiscovery) error {
+	watcher, err := fsnotify.NewWatcher()
```

Comment:
Using fsnotify to watch for Kubernetes secret changes might not always work, due to it using symlinks (see #13621 for how to reproduce). Have you checked if you could use the `fswatcher` package instead?

https://github.com/cilium/cilium/blob/dfa6b157e8e9484c65fd938b2b45a4f5f50c61f9/pkg/crypto/certloader/fswatcher/fswatcher.go#L25-L50